mmaprototype: fix isConstraintSatisfied

wenyihu6 opened the pull request:
Epic: CRDB-55052 
Release note: none

---
**mmaprototype: add test case for isConstraintSatisfied**

This commit adds a test case for isConstraintSatisfied to show that it produces
unintended results with voter constraints, as it counts both voters and
non-voters when evaluating constraint satisfaction.

---
**mmaprototype: fix isConstraintSatisfied**

Previously, we were incorrectly including both voterIndex and nonVoterIndex in
satisfiedByReplica for voter constraints's isConstraintSatisfied. This commit
fixes isConstraintSatisfied by updating it to check only voters when evaluating
voter constraints.

---
**mmaprototype: add test case for over-assigning nonvoters**

This commit adds a test case for range analyzed constraint to
illustrate the reason why isConstraintSatisfied needs to
include both voter and nonvoter for voter constraints.

---
**mmaprototype: fix voter constraint check**

finishInit currently prioritizes assigning stores that satisfy only one
constraint as an optimization. However, this can be suboptimal when the replica
is a non-voter, and we are satisfying voter constraints; in that case, we should
delay assigning the non-voter to give voters a chance to satisfy the constraint
first, even if those voters match multiple constraints.

wenyihu6 committed:
mmaprototype: add test case for isConstraintSatisfied

This commit adds a test case for isConstraintSatisfied to show that it produces
unintended results with voter constraints, as it counts both voters and
non-voters when evaluating constraint satisfaction.

wenyihu6 committed:
mmaprototype: fix isConstraintSatisfied

Previously, we were incorrectly including both voterIndex and nonVoterIndex in
satisfiedByReplica for voter constraints's isConstraintSatisfied. This commit
fixes isConstraintSatisfied by updating it to check only voters when evaluating
voter constraints.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159228)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1481 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgDuijU1ywthtdxVWkb:-OgDuijU1ywthtdxVWkc:b7mbhjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de1beb3199b2e850cc3046ab157420f57ce0a6f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1481)):*
> ```Go
> 			// For voter constraints, only voters can satisfy them.
> 			return len(ac.satisfiedByReplica[voterIndex][constraintIndex]) >= int(ac.constraints[constraintIndex].numReplicas)
> 		}
> ```

The existing behavior was deliberate to avoid over-satisfying with non-voters.

For instance, consider a variant of your example:
[+region=a]:1, [+zone=b1]:1, [+region=b]:1

and two non-voters with attributes region=a,zone=b1 and one voter in region=b. The existing voter part is fine. But if we only count voters in the `isConstraintSatisfied` step we will assign both non-voters to [+region=a] instead of spreading them out between [+region=a], [+zone=b1].

That is, the phase 2 is trying to not over-satisfy to ensure that every constraint gets a chance to be satisfied. And the reason that it is fine to count both voters and non-voters is that it first iterates over all voters (see below).

What I don't understand (glancing quickly at the code) on why this is not working as intended in the test case you added. The phase 2 loop is first looking at all the stores corresponding to the `voterIndex`. So for the `+region=a:2` voter constraint, when it does `for i := range buf.replicaConstraintIndices[kind]` where kind=voterIndex it should encounter both s10, s12, and only later when kind=nonVoterIndex it should encounter s11. So it should satisfy that constraint with both voters. I suspect there is a bug in that loop structure and not in the existing behavior of `isConstraintSatisfied`.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1481 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgDuijU1ywthtdxVWkb:-OgFP9lo2DVPTin-UCdq:b-5qda0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de1beb3199b2e850cc3046ab157420f57ce0a6f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1481)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The existing behavior was deliberate to avoid over-satisfying with non-voters.

For instance, consider a variant of your example:
[+region=a]:1, [+zone=b1]:1, [+region=b]:1

and two non-voters with attributes region=a,zone=b1 and one voter in region=b. The existing voter part is fine. But if we only count voters in the `isConstraintSatisfied` step we will assign both non-voters to [+region=a] instead of spreading them out between [+region=a], [+zone=b1].

That is, the phase 2 is trying to not over-satisfy to ensure that every constraint gets a chance to be satisfied. And the reason that it is fine to count both voters and non-voters is that it first iterates over all voters (see below).

What I don't understand (glancing quickly at the code) on why this is not working as intended in the test case you added. The phase 2 loop is first looking at all the stores corresponding to the `voterIndex`. So for the `+region=a:2` voter constraint, when it does `for i := range buf.replicaConstraintIndices[kind]` where kind=voterIndex it should encounter both s10, s12, and only later when kind=nonVoterIndex it should encounter s11. So it should satisfy that constraint with both voters. I suspect there is a bug in that loop structure and not in the existing behavior of `isConstraintSatisfied`.
</blockquote></details>

I think in my example we never entered the loop satisfaction because s10 and s12 each satisfy only one constraint - `[+region=a]` so we assigned a constraint for them early https://github.com/cockroachdb/cockroach/blob/b12dce952a37c83c65230c6c4673ea3e089c9c79/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1462, and since we count both non voter and voter for the constraint, this constraint was considered satisfied at the start of the loop https://github.com/cockroachdb/cockroach/blob/b12dce952a37c83c65230c6c4673ea3e089c9c79/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1493-L1495.‚‚

 I see what you mean about the need to avoid over-satisfying with non-voters. Need to think about what the right fix is here. 


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1481 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgDuijU1ywthtdxVWkb:-OgF_JPB6dPCZfCCL_MQ:bf64ryh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de1beb3199b2e850cc3046ab157420f57ce0a6f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1481)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I think in my example we never entered the loop satisfaction because s10 and s12 each satisfy only one constraint - `[+region=a]` so we assigned a constraint for them early https://github.com/cockroachdb/cockroach/blob/b12dce952a37c83c65230c6c4673ea3e089c9c79/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1462, and since we count both non voter and voter for the constraint, this constraint was considered satisfied at the start of the loop https://github.com/cockroachdb/cockroach/blob/b12dce952a37c83c65230c6c4673ea3e089c9c79/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1493-L1495.‚‚

 I see what you mean about the need to avoid over-satisfying with non-voters. Need to think about what the right fix is here.
</blockquote></details>

I see.

I think the right fix is:
```
} else if n == 1 && (!isVoterConstraints || kind == voterIndex) {
```

The point was to place the ones that can only satisfy one up front. But if it is a non-voter and we are looking at a voter constraint, it is preferable to delay it, since we would like to give an opportunity for voters to satisfy the constraint, even if a voter can satisfy multiple.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: add test case for over-assigning nonvoters

This commit adds a test case for range analyzed constraint to
illustrate the reason why isConstraintSatisfied needs to
include both voter and nonvoter for voter constraints.

wenyihu6 committed:
mmaprototype: fix voter constraint check

finishInit currently prioritizes assigning stores that satisfy only one
constraint as an optimization. However, this can be suboptimal when the replica
is a non-voter, and we are satisfying voter constraints; in that case, we should
delay assigning the non-voter to give voters a chance to satisfy the constraint
first, even if those voters match multiple constraints.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1481 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgDuijU1ywthtdxVWkb:-OgFTTdL5oQZuMvfo2Ig:b5eg7le) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de1beb3199b2e850cc3046ab157420f57ce0a6f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1481)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I see.

I think the right fix is:
```
} else if n == 1 && (!isVoterConstraints || kind == voterIndex) {
```

The point was to place the ones that can only satisfy one up front. But if it is a non-voter and we are looking at a voter constraint, it is preferable to delay it, since we would like to give an opportunity for voters to satisfy the constraint, even if a voter can satisfy multiple.
</blockquote></details>

To confirm my understanding: in phase two, we iterate over constraints and, for each constraint, consider existing voters first and existing non-voters second. That means while we iterate over all voters to populate `ac.satisfiedByReplica[voterIndex][constraintIndex]`, we should not have assigned anything to `ac.satisfiedByReplica[nonVoterIndex][constraintIndex]` yet.

So I could add an assertion like this right? 
```
	isConstraintSatisfied := func(constraintIndex int, kind replicaKindIndex) bool {
		if isVoterConstraints && kind == voterIndex {
			if len(ac.satisfiedByReplica[nonVoterIndex][constraintIndex]) != 0 {
				panic("hey")
			}
		}
        }
```


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> @sumeerbhola reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1481 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgDuijU1ywthtdxVWkb:-OgNveGj58Cd-zmWNJ8N:b-56gqsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de1beb3199b2e850cc3046ab157420f57ce0a6f/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1481)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

To confirm my understanding: in phase two, we iterate over constraints and, for each constraint, consider existing voters first and existing non-voters second. That means while we iterate over all voters to populate `ac.satisfiedByReplica[voterIndex][constraintIndex]`, we should not have assigned anything to `ac.satisfiedByReplica[nonVoterIndex][constraintIndex]` yet.

So I could add an assertion like this right? 
```
	isConstraintSatisfied := func(constraintIndex int, kind replicaKindIndex) bool {
		if isVoterConstraints && kind == voterIndex {
			if len(ac.satisfiedByReplica[nonVoterIndex][constraintIndex]) != 0 {
				panic("hey")
			}
		}
        }
```
</blockquote></details>

Yes, that assertion looks correct.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1446 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNvomy-WOLVEQuRddU:-OgNvomy-WOLVEQuRddV:b-hhg83o) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1446)):*
> ```Go
> 	// constraint and ac.satisfiedByReplica[kind][i] for stores that satisfy
> 	// exactly one constraint. Since we will be assigning at least one
> 	// constraint to each store, these stores are unambiguous.
> ```

The last couple of sentences need updating. One possibility is to add something like

// An exception to this is when we are considering voter constraints. We delay assigning non voters that satisfy only one constraint until phase 2, since we want to first assign the voters that satisfy multiple constraints. That is, we want to avoid a situation where we unnecessarily need to elevate a non-voter to voter.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1486 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNwRva4_lyhlZpVCQs:-OgNwRva4_lyhlZpVCQt:bmlthva) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1486)):*
> ```Go
> 	// Example:
> 	// constraints [+region=a]:1, [+zone=b1]:1, [+region=b]:1 and two non-voters
> 	// matching both region=a and zone=b1, counting only voters would cause both
> ```

nit: missing period before "counting only voters ..."
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1488 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNwYhd38l5Uxo5En9Z:-OgNwYhd38l5Uxo5En9_:bxt58ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1488)):*
> ```Go
> 	// matching both region=a and zone=b1, counting only voters would cause both
> 	// non-voters to be assigned to [+region=a] instead of spreading them across
> 	// constraints. See range_analyzed_constraints for more details.
> ```

nit: see testdata/range_analyzed_constraints for examples illustrating this behavior.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints` line 674 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNxP4h5gCynOXXP5Os:-OgNxP4iAt9RVMno8nar:bnyxyip) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints#L674)):*
> ```
> # voters for voter constraints.
> #
> # In analyzedConstraints.initialize, isConstraintSatisfied determines if a
> ```

This comment is stale
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints` line 745 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNxVs4COhJgyitZZgF:-OgNxVs4COhJgyitZZgG:bmsyk0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints#L745)):*
> ```
> # Test: Non-voter spreading for all-replica constraints.
> #
> # This verifies that the fix for voter constraints (only count voters) does NOT
> ```

We should try to avoid writing comments like "fix for voter constraints" (sometimes a pointer to a github issue that has a good example is ok), since a typical code reader sees the code as it currently is, and not the history.

Also, I'm confused on why it says "all-replica constraints" where the spreading behavior we are testing here is for non-voters on a voter constraint.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1446 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNvomy-WOLVEQuRddU:-OgY9Iii-Y3PvCO_Vu1q:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1446)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The last couple of sentences need updating. One possibility is to add something like

// An exception to this is when we are considering voter constraints. We delay assigning non voters that satisfy only one constraint until phase 2, since we want to first assign the voters that satisfy multiple constraints. That is, we want to avoid a situation where we unnecessarily need to elevate a non-voter to voter.
</blockquote></details>

Reworded.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1486 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNwRva4_lyhlZpVCQs:-OgY9Jmz5zj2G-kj7n2G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1486)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: missing period before "counting only voters ..."
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/constraint.go` line 1488 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNwYhd38l5Uxo5En9Z:-OgY9LmR3BtRHFhE1YVW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/constraint.go#L1488)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: see testdata/range_analyzed_constraints for examples illustrating this behavior.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints` line 674 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNxP4h5gCynOXXP5Os:-OgY9NrB-JjF9wPHlw97:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints#L674)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This comment is stale
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints` line 745 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-OgNxVs4COhJgyitZZgF:-OgY9PEQ9xcgZn7-_-8a:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/415553fb889db6259f4aadd788f975bbeeb176ea/pkg/kv/kvserver/allocator/mmaprototype/testdata/range_analyzed_constraints#L745)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We should try to avoid writing comments like "fix for voter constraints" (sometimes a pointer to a github issue that has a good example is ok), since a typical code reader sees the code as it currently is, and not the history.

Also, I'm confused on why it says "all-replica constraints" where the spreading behavior we are testing here is for non-voters on a voter constraint.
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159228#-:-OgYLVl65Egz7SEZ8LzC:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)



<!-- Sent from Reviewable.io -->
